File: app/models/quantity.rb
# A quantity of a food. Has a value, sometimes a unit.
class Quantity
  attr_reader :unit, :food, :raw, :grams, :ml, :total

  def initialize(food)
    @food = food
  end

  # Only grams, ml and total are good. The others are pretty much deprecated
  def set_from_grams(grams)
    @grams = grams
    if grams.nil?
      @ml = nil
      @total = nil
      return self
    end
    @ml = @grams / @food.density if @food && !@food.density.nil?
    @total = @grams / @food.unit_weight if !@grams.nil? && @food && !@food.unit_weight.nil?
    self
  end

  def set_from_value_and_unit(qty, unit)
    factor = unit && unit.value ? unit.value : 1.0
    @unit = unit
    if unit.nil? || unit.is_unitary
      @total = qty * factor
      @grams = @total * @food.unit_weight if @food && !@food.unit_weight.nil?
      @ml = @grams / @food.density if !@grams.nil? && !@food.density.nil?
    elsif unit.is_volume
      @ml = qty * factor
      @grams = @ml * @food.density if @food && !@food.density.nil?
      @total = @grams / @food.unit_weight if !@grams.nil? && @food && !@food.unit_weight.nil?
    else
      @grams = qty * factor
      @ml = @grams / @food.density if @food && !@food.density.nil?
      @toal = @grams / @food.unit_weight if !@grams.nil? && @food && !@food.unit_weight.nil?
    end
    self
  end

  def self.parse_float(_qty_s)
    return nil if _qty_s.blank?
    qty_s = _qty_s.strip
    #qty_s = _raw[/^\d+[,./]\d+/]
    if qty_s.include?("/")
      if qty_s.include?(" ")
        whole, fraction = qty_s.split(' ')
        whole.to_i + fraction.to_r.to_f
      else
        qty_s.to_r.to_f
      end
    else
      qty_s.to_f
    end
  end

  def set_from_raw(raw)
    return nil if raw.nil?
    @raw = raw
    qty_s = raw[/^\d+( \d)?([,.\/]\d+)?/]
    qty = Quantity.parse_float(qty_s)
    return nil if qty.nil?
    unit_s = raw[qty_s.length..-1].strip
    set_from_value_and_unit(qty, Unit.find_by(name: unit_s))
  end

  def to_raw
    return "#{@total}" if @unit.nil?
    return "#{@total} #{@unit.name}" if @unit.is_unitary
    return "#{@ml/(@unit.value || 1.0)} #{@unit.name}" if @unit.is_volume
    return "#{@grams/(@unit.value || 1.0)} #{@unit.name}"
  end

  #def unit_quantity
  #  return @unitary_qty if @unit.nil? || @unit.is_unitary
  #  return @volume if @unit.is_volume
  #  return @weight
  ##end

  def self.ratio(from, to)
    raise "Invalid quantity ratio" unless from.food == to.food
    from.grams / to.grams
  end

  def *(scalar)
    @grams *= scalar if @grams
    @total *= scalar if @total
    @ml *= scalar if @ml
    self
  end
end

File: app/javascript/models/quantity.js
import {extractNumberAtBeginning} from "../utils"

// A quantity is a number with an optional label.
// Used for ingredients and for servings.
// Examples:
// 12 muffins
// 1/2 t
// 250 mL
// 4
export default class Quantity {

  constructor(args = {}) {
    this.raw = args.raw
    this.nb = null
    this.label = null

    this.ml = null
    this.grams = null
    this.total = null

    this.unit = null

    if (this.raw != null) {
      let s = extractNumberAtBeginning(this.raw)
      if (s != null) {
        let qty_s = s
        this.label = this.raw.substr(qty_s.length).trim()
        this.nb = Quantity.parseFloatOrFraction(qty_s)
        this.unit = gon.units ? gon.units.find(unit => unit.name == this.label) : null
      }
    }
  }

  // TODO: Try to parse complete ingredient separated by de or d'
  // Then, try to parse quantity. Check if label is a unit. Check if label is a food.
  static parseQuantityAndFoodName(raw) {
    const separators = [";", "de", "d'"]
    for (let i = 0; i < separators.length; i++) {
      if (raw.includes(separators[i])) {
        const s = raw.split(separators[i])
        let rawQty = s[0].trim()
        let foodName = s[1].trim()
        let qty = new Quantity({raw: rawQty})
        return [qty, foodName]
      }
    }
    let qty = new Quantity({raw: raw})
    if (qty.unit) {return [qty, null] }
    let foodName = qty.label
    qty.label = null
    return [qty, foodName]
  }

  static parseFractionFloat(str) {
    var split = str.split('/')
    return split[0]/split[1]
  }

  static parseFloatOrFraction(str) {

    if (!str) {return null;}
    str = str.replace(',', '.') // FIXME: Inefficient way of allowing commas instead of dots. Also I believe some country use 2,500,500.00 which does not work with this, but this should not be a problem with recipes...
    var qty_s = str.trim()
    if (qty_s.includes("/")) {
      if (qty_s.includes(" ")) {
        var s = qty_s.split(' ')
        var whole = s[0]
        var fraction = s[1]
        return parseInt(whole, 10) + Quantity.parseFractionFloat(fraction)
      } else {
        return Quantity.parseFractionFloat(qty_s)
      }
    } else {
      return parseFloat(qty_s)
    }
  }

  pretty() {
    if (!this.label) {
      return this.nb.toString()
    } else {
      return this.raw
    }
  }
}
